pkg/ostree/app.js

```diff
@@ -0,0 +1,328 @@
+define([
+    "jquery",
+    "base1/cockpit",
+    "base1/angular",
+    "base1/moment",
+    "ostree/client",
+], function($, cockpit, angular, moment, client) {
+    'use strict';
+
+    var _ = cockpit.gettext;
+    var phantom_checkpoint = phantom_checkpoint || function () { };
+
+    function track(item) {
+        if (!item)
+            return;
+
+        var key = item.osname.v;
+        if (item.id)
+            key = key + item.id.v;
+
+        if (item.checksum)
+            key = key + item.checksum.v;
+
+        return key;
+    }
+
+    function notify_result(promise, scope) {
+        promise
+            .progress(function(msg) {
+                scope.$applyAsync(function() {
+                    scope.progressMsg = msg;
+                });
+            })
+            .fail(function(ex) {
+                scope.$applyAsync(function() {
+                    scope.error = cockpit.format(cockpit.message(ex));
+                });
+            })
+            .always(function(ex) {
+                scope.$applyAsync(function() {
+                    scope.progressMsg = null;
+                });
+            });
+    }
+
+
+    return angular.module('ostree', [
+            'ngRoute',
+        ])
+        .config([
+            '$routeProvider',
+             function($routeProvider) {
+                $routeProvider.when('/', {
+                    templateUrl: 'index.html',
+                    controller: 'indexController'
+                });
+        }])
+
+        .controller('mainController', [
+            '$scope',
+            '$timeout',
+            'ostreeClient',
+            function($scope, $timeout, client) {
+                function show_failure(ex) {
+                    var message = null;
+                    if (ex.problem === "access-denied")
+                        message = _("Not authorized to update software on this system");
+                    else
+                        message = cockpit.message(ex);
+
+                    $scope.curtains = { state: 'failed',
+                                        failure: ex,
+                                        message: message };
+                    $scope.$digest();
+                    phantom_checkpoint();
```

Comment:
Suggest implementing a helper function for doing these two together. Either that, or figure out a way to call phantom_checkpoint() globally for any digest.